### PR TITLE
🚧Add README pre-processing feature

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

File: CICD/CICD.Release.Preview.cs [hunk 1]
```diff
@@ -77,6 +77,14 @@ public partial class CICD // Release.Preview
...
 
+                // If README pre-processing should be performed, process readme before packaging
+                if (PreProcessReadMe)
+                {
+                    var readmeService = App.Container.GetInstance<IReadmeService>();
+
+                    readmeService.RunPreProcessing();
+                }
+
                 // Create the nuget package to deploy
```
Comment (kselena) [inline CICD/CICD.Release.Preview.cs:88]: Refactor nuget to NuGet (it's everywhere)
```suggestion
                // Create the Nuget package to deploy
```

File: CICD/CICD.Release.Preview.cs [hunk 2]
```diff
@@ -77,6 +77,14 @@ public partial class CICD // Release.Preview
                     $"{Environment.NewLine}{ConsoleTab}To view the milestone, go here 👉🏼 {updatedMilestone.HtmlUrl}{Environment.NewLine}";
                 Log.Information(updateMsg);
 
+                // If README pre-processing should be performed, process readme before packaging
```
Comment (kselena) [inline CICD/CICD.Release.Preview.cs:80]: I know these are just comments, but you need to make a decision on README or readme or Readme.  I say README is more prominent, so I would refactor to README.
```suggestion
                // If README pre-processing should be performed, process README before packaging
```

File: CICD/Services/Interfaces/IFindDirService.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+namespace CICDSystem.Services.Interfaces;
+
+/// <summary>
+/// Provides directory search functionality.
+/// </summary>
+public interface IFindDirService
+{
+    /// <summary>
+    /// Walks up one directory at a time starting at the given <paramref name="startPath"/>
+    /// until a directory with a name that matches the given <paramref name="dirNameToFind"/>.
```
Comment (kselena) [inline CICD/Services/Interfaces/IFindDirService.cs:14]: Grammar
```suggestion
    /// Walks up one directory at a time starting at the given <paramref name="startPath"/>
    /// until a directory with a name that matches the given <paramref name="dirNameToFind"/> is found.
```

File: CICD/Services/Interfaces/IFindDirService.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    /// </summary>
+    /// <param name="startPath">The directory path to start the search from.</param>
```
Comment (kselena) [inline CICD/Services/Interfaces/IFindDirService.cs:16]: Grammar
```suggestion
    /// <param name="startPath">The directory path to start the search.</param>
```

File: CICD/Services/ReadmeService.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+namespace CICDSystem.Services;
+
+/// <inheritdoc/>
+internal sealed class ReadmeService : IReadmeService
+{
+    private const string MultiCommentStart = "<!--";
+    private const string MultiCommentStop = "-->";
+    private const string GitHubDirName = ".github";
+    private const string ReadMeFileName = "README.md";
+    private const string ProcessingCommentStart = "<!--PRE-PROCESSING-COMMENT-START-->"; // TODO: setup and str interpolation
```
Comment (kselena) [inline CICD/Services/ReadmeService.cs:22]: Do you need this TODO?
```suggestion
    private const string ProcessingCommentStart = "<!--PRE-PROCESSING-COMMENT-START-->"; // TODO: setup and str interpolation
```

File: CICD/Services/ReadmeService.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        this.baseDirPath = path.GetDirectoryName(this.baseDirPath)
+            .Replace('\\', '/');
+    }
+
+    /// <inheritdoc/>
+    /// <exception cref="FileNotFoundException">
+    ///     Thrown when the readme file is not found.
+    /// </exception>
+    /// <exception cref="ReadmeProcessingException">
+    ///     Thrown when there is an issue with processing the file.
```
Comment (kselena) [inline CICD/Services/ReadmeService.cs:61]: Grammar
```suggestion
    ///     Thrown when there is an issue processing the file.
```

File: CICD/Services/ReadmeService.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        this.baseDirPath = findDirService.FindDescendentDir(directory.GetCurrentDirectory(), GitHubDirName)
+            .Replace('\\', '/');
+
+        this.baseDirPath = path.GetDirectoryName(this.baseDirPath)
+            .Replace('\\', '/');
+    }
+
+    /// <inheritdoc/>
+    /// <exception cref="FileNotFoundException">
+    ///     Thrown when the readme file is not found.
```
Comment (kselena) [inline CICD/Services/ReadmeService.cs:58]: Refactor readme to README
```suggestion
    ///     Thrown when the README file is not found.
```

File: Testing/CICDTests/Services/FindDirServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FindDirServiceTests"/> class.
+    /// </summary>
+    public FindDirServiceTests()
+    {
+        this.mockDirectory = new Mock<IDirectory>();
+        this.mockDirectory.Setup(m => m.GetCurrentDirectory()).Returns(DebugDir);
+
+        // Starting at the execution dir
```
Comment (kselena) [inline Testing/CICDTests/Services/FindDirServiceTests.cs:35]: I know this is a comment, but to be consistent, refactor dir to directory.  
```suggestion
        // Starting at the execution directory
```

File: Testing/CICDTests/Services/ReadmeServiceTests.cs [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+        this.mockFile.Setup(m => m.ReadAllLines(It.IsAny<string>()))
+            .Returns(fileContent);
+
+        var sut = CreateSystemUnderTest();
+
+        // Act
+        var act = () => sut.RunPreProcessing();
+
+        // Assert
+        // TODO: need to use different exception
```
Comment (kselena) [inline Testing/CICDTests/Services/ReadmeServiceTests.cs:199]: Is this TODO still needed? 
```suggestion
        // TODO: need to use different exception
```

File: Testing/CICDTests/Services/ReadmeServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
```
Comment (kselena) [inline Testing/CICDTests/Services/ReadmeServiceTests.cs:231]: Is this TODO still needed? 
```suggestion
        // TODO: need to use different exception
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by kselena — DISMISSED** [review]

**Review by kselena — APPROVED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]
